src/SDiagonal.jl

```diff
@@ -0,0 +1,109 @@
+# Originally contributed by D. Getz (https://github.com/getzdan), M. Schauer
+# at https://github.com/mschauer/Bridge.jl under MIT License
+
+import Base: getindex,setindex!,==,-,+,*,/,\,transpose,ctranspose,convert, size, abs, real, imag, conj, eye, inv
+import Base.LinAlg: ishermitian, issymmetric, isposdef, factorize, diag, trace, det, logdet, expm, logm, sqrtm
+
+@generated function scalem{M, N}(a::StaticMatrix{M,N}, b::StaticVector{N})
+    expr = vec([:(a[$j,$i]*b[$i]) for j=1:M, i=1:N])
+    :(let val1 = ($(expr[1])); similar_type(SMatrix{M,N},typeof(val1))(val1, $(expr[2:end]...)); end)
+end
+@generated function scalem{M, N}(a::StaticVector{M}, b::StaticMatrix{M, N})
+    expr = vec([:(b[$j,$i]*a[$j]) for j=1:M, i=1:N])
+    :(let val1 = ($(expr[1])); similar_type(SMatrix{M,N},typeof(val1))(val1, $(expr[2:end]...)); end)
+end
+
+struct SDiagonal{N,T} <: StaticMatrix{N, N, T}
+    diag::SVector{N,T}
+    SDiagonal{N,T}(diag::SVector{N,T}) where {N,T} = new(diag)
+end    
+diagtype{N,T}(::Type{SDiagonal{N,T}}) = SVector{N,T}
+diagtype{N}(::Type{SDiagonal{N}}) = SVector{N}
+diagtype(::Type{SDiagonal}) = SVector
+
+# this is to deal with convert.jl
+@inline (::Type{SD})(a::AbstractVector) where {SD <: SDiagonal} = SDiagonal(convert(diagtype(SD), a))
+@inline (::Type{SD})(a::Tuple) where {SD <: SDiagonal} = SDiagonal(convert(diagtype(SD), a))
+@inline (::Type{SDiagonal}){N,T}(a::SVector{N,T}) = SDiagonal{N,T}(a) 
+
+@generated function SDiagonal{N,T}(a::StaticMatrix{N,N,T})
+    expr = [:(a[$i,$i]) for i=1:N]
+    :(SDiagonal{N,T}($(expr...)))
+end
+
+convert{N,T}(::Type{SDiagonal{N,T}}, D::SDiagonal{N,T}) = D
+convert{N,T}(::Type{SDiagonal{N,T}}, D::SDiagonal{N}) = SDiagonal{N,T}(convert(SVector{N,T}, D.diag))
+
+Base.@propagate_inbounds function getindex{N,T}(D::SDiagonal{N,T}, i::Int, j::Int)  
```

Comment:
`@propagate_inbounds` is needed if you want to defer bounds checking to an inner function. This should be simply `@inline` since you are doing your own `@boundscheck` and `@inbounds` here.